accepted/2021/preview-features/preview-features.md

```diff
@@ -0,0 +1,418 @@
+  </PropertyGroup>
+
+</Project>
+```
+
+The upside is that one can be explicit about which preview feature one wants to
+turn on (and thus depend on).
+
+The downsides were:
+
+1. It's a more complicated concept for the user to understand and very different
+   from how previews of the .NET platform itself work. This might increase the
+   barrier of entry, which we don't want.
+
+2. The test matrix for the product gets much more complicated as the combinations
+   of on/states explode quickly.
+
+3. Not all preview features have a convenient entry point (language syntax,
+   API). Having a global preview mode increases airtime for features that
+   otherwise wouldn't have been turned on.
+
+4. We're unlikely to ship a grab bag of preview features and instead are more
+   likely to ship a cohesive feature set that in its totality is still in
+   preview, such as static interface members, which includes runtime work,
+   language work, and compiler work. Said differently, having a single preview
+   mode forces us to jointly coordinate preview features across all layers.
+
+### Why are we recording whether an assembly used preview features?
+
+For two reasons:
+
+1. It can be used for auditing
+2. It is used to enforce transitivity. An assembly that doesn't use preview
+   features shouldn't depend on assemblies that do. It doesn't matter whether
+   this binary was a project-to-project reference, a NuGet package, or a raw
+   reference to some checked in file -- we only have to look at the assembly
+   level attribute.
+
+### What happens when a preview feature is released?
+
+Several things will happen:
+
+1. We remove the `[RequiresPreviewFeatures]` attribution from any APIs
+2. We remove any guards from the runtime/compilers that check for the preview
+   flag (of course, compilers are likely still checking language versions, but
+   that's separate)
+
+If someone shipped a binary to nuget.org that depends on a preview feature,
+consuming it from a .NET version where the feature was released will still
+generate an error unless the consuming project has preview features turned on.
+
+### Can 3rd parties introduce their own preview APIs?
+
+In this design the system has no way to prevent this and it will in fact just
+work.
+
+### Why did we use the term *preview*?
+
+Originally I used the term *experimental* but we decided to change it to
+*preview* for these reasons:
+
+* The quality bar and connotation of preview features is virtually identical to
+  preview builds of the platform. The only different is that an RTM build
+  additionally contains features that are still in the preview state.
+* The compiler has a language version `preview` that has the same connotation
+  that we use here.
+* We generally used the term *experimental* where probability of shipping is
+  <50%, e.g. in `runtimelabs`
+* Overall, the goal is draw more people in and motivate them to try the
+  features. The term *preview* seems more helpful than *experimental*.
```

Comment:
What about preview SDK features, is that out of scope? Eg a dotnet.exe feature. It presumably could not be enabled at project level, in general.

Reply:
I consider that out of scope because it's not clear to me that this maps well to the project level.